pkg/sql/fingerprint/fingerprint_job.go

```diff
@@ -0,0 +1,318 @@
+		if err != nil {
+			return err
+		}
+		if !ok {
+			// No checkpoint exists yet.
+			return nil
+		}
+
+		found = true
+		state.fingerprint = hash
+		done, ok, err := jobfrontier.Get(ctx, txn, p.id, frontier)
+		if err != nil {
+			return err
+		}
+		if ok {
+			state.done = done
+		}
+		return nil
+	})
+
+	if err != nil {
+		return checkpointState{}, false, err
+	}
+
+	return state, found, nil
+}
+
+func (p *persist) store(ctx context.Context, state checkpointState, frac float64) error {
+	return p.execCtx.ExecCfg().InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		infoStorage := jobs.InfoStorageForJob(txn, p.id)
+		if err := infoStorage.WriteUint64(ctx, infoKey, state.fingerprint); err != nil {
+			return errors.Wrap(err, "failed to store fingerprint hash")
+		}
+
+		if state.done != nil && !state.done.Frontier().IsEmpty() {
+			if err := jobfrontier.Store(ctx, txn, p.id, frontier, state.done); err != nil {
+				return errors.Wrap(err, "failed to store done frontier")
+			}
+		} else {
+			if err := jobfrontier.Delete(ctx, txn, p.id, frontier); err != nil {
+				return errors.Wrap(err, "failed to delete done frontier")
+			}
+		}
+		if err := jobs.ProgressStorage(p.id).Set(ctx, txn, frac, hlc.Timestamp{}); err != nil {
+			return errors.Wrap(err, "failed to update job progress")
+		}
+		return nil
+	})
+}
+
+type partitionSpans struct {
+	sql.JobExecContext
+}
+
+var _ partitioner = partitionSpans{}
+
+func (p partitionSpans) partition(
+	ctx context.Context, spans []roachpb.Span,
+) ([]sql.SpanPartition, error) {
+	dsp := p.DistSQLPlanner()
+	planCtx, _, err := dsp.SetupAllNodesPlanningWithOracle(
+		ctx, p.ExtendedEvalContext(),
+		p.ExecCfg(),
+		kvfollowerreadsccl.NewBulkOracle(
+			dsp.ReplicaOracleConfig(p.ExecCfg().Locality), roachpb.Locality{}, kvfollowerreadsccl.StreakConfig{},
+		),
+		roachpb.Locality{},
+	)
+	if err != nil {
+		return nil, err
+	}
+	spanPartitions, err := p.DistSQLPlanner().PartitionSpans(ctx, planCtx, spans, sql.PartitionSpansBoundDefault)
+	if err != nil {
+		return nil, err
+	}
+	return spanPartitions, nil
+}
+
+type kvFingerprinter struct {
+	asOf     hlc.Timestamp
+	start    hlc.Timestamp
+	stripped bool
+	sender   kv.Sender
+}
+
+var _ spanFingerprinter = kvFingerprinter{}.fingerprintSpan
+
+func (k kvFingerprinter) fingerprintSpan(
+	ctx context.Context, span roachpb.Span,
+) (spanFingerprintResult, error) {
+	res := spanFingerprintResult{span: span}
+
+	header := kvpb.Header{Timestamp: k.asOf, ReturnElasticCPUResumeSpans: true}
+
+	admissionHeader := kvpb.AdmissionHeader{
+		Priority:                 int32(admissionpb.BulkNormalPri),
+		CreateTime:               timeutil.Now().UnixNano(),
+		Source:                   kvpb.AdmissionHeader_FROM_SQL,
+		NoMemoryReservedAtSource: true,
+	}
+	for len(span.Key) != 0 {
+		req := &kvpb.ExportRequest{
+			RequestHeader:      kvpb.RequestHeader{Key: span.Key, EndKey: span.EndKey},
+			StartTime:          k.start,
```

Comment:
without revision history, we should not be setting the start time or keeping in the job details, yeah?